### ISSUE TITLE
Sometimes the generated key can't be imported by Go crypto/tls

### ISSUE THREAD

**Issue by Foosvald** [digitalbazaar/forge#431]
Thanks for a good lib!

I have a small problem though, which I also posted at https://github.com/golang/go/issues/17279

We are using version of Node-forge 0.6.42, and get a problem both in OSX and Linux (running in Node). 

This is the code we use to generate a key pair:

```
'use strict'

const fs = require('fs')
const forge = require('node-forge')

const caCert = fs.readFileSync(__dirname + '/fixtures/ca_test.pem')
const caKey = fs.readFileSync(__dirname + '/fixtures/ca_test.key')

let cert = generateServerCertificate(
    caKey, caCert, 'etcd', ['127.0.0.1'])

fs.writeFileSync(__dirname + '/etcd.pem', cert.cert)
fs.writeFileSync(__dirname + '/etcd.key', cert.key)

function generateServerCertificate(caKeyString, caCertString, serviceInstance, ips) {
    var caKey = forge.pki.privateKeyFromPem(caKeyString)
    var caCert = forge.pki.certificateFromPem(caCertString)

    var keys = forge.pki.rsa.generateKeyPair(2048)
    var cert = forge.pki.createCertificate()
    cert.publicKey = keys.publicKey
    cert.serialNumber = '01'
    cert.validity.notBefore = new Date()
    cert.validity.notAfter = new Date()
    cert.validity.notAfter.setFullYear(cert.validity.notBefore.getFullYear() + 1)

    var attrsSubject = [
        {
            name: 'commonName',
            value: 'none'
        },
        {
            name: 'organizationName',
            value: 'Our Org'
        },
        {
            shortName: 'OU',
            value: serviceInstance
        }
    ]

    cert.setSubject(attrsSubject)
    cert.setIssuer(caCert.subject.attributes)

    cert.setExtensions([
        {
            name: 'subjectAltName',
            altNames: ips.map(ip => {
                return { type: 7, ip: ip }
            }),
        }])

    cert.sign(caKey, forge.md.sha256.create())

    var pem = {
        key: forge.pki.privateKeyToPem(keys.privateKey),
        cert: forge.pki.certificateToPem(cert)
    }

    return pem
}
```
And then we import it in Go using 
```
package main

import (
    "crypto/tls"
    "log"
)

func main() {
    _, err := tls.LoadX509KeyPair("broken.pem", "broken.key")
    if(err != nil) {
        log.Fatal(err)
    }
    log.Print("Go loaded key successfully")
}
```
And in most cases the key is successfully imported by Go, but sometimes it fails with the following error message: `tls: failed to parse private key` which apparently deeper down in the Go crypto/tls module is caused by `asn1: structure error: integer not minimally-encoded`.

I've attached both a pair of working keys and a pair of broken keys which have been generated in the same way.
[keys.zip](https://github.com/digitalbazaar/forge/files/502463/keys.zip)

Is this a problem with how we generate and save the key, or is it a problem deeper down in the code?

**Comment by dlongley** [comment]
@Foosvald,

I took a look at the broken key and the working key. The working key is in PKCS#1 format and the broken key is in PKCS#8 format. So I can't figure out how they could possibly be generated from the same code. You must call different methods in forge to get different formatted keys. The code sample you pasted would generate a PKCS#1 key, not a PKCS#8 key. It looks like Go can't parse the PKCS#8 formatted key.

To produce a PKCS#8 formatted key in forge you must do this:

```
// convert a Forge private key to an ASN.1 RSAPrivateKey
var rsaPrivateKey = pki.privateKeyToAsn1(privateKey);

// wrap an RSAPrivateKey ASN.1 object in a PKCS#8 ASN.1 PrivateKeyInfo
var privateKeyInfo = pki.wrapRsaPrivateKey(rsaPrivateKey);

// convert a PKCS#8 ASN.1 PrivateKeyInfo to PEM
var pem = pki.privateKeyInfoToPem(privateKeyInfo);
```

**Comment by Foosvald** [comment]
Thanks for looking into it @dlongley!

I'm running the same code each time, which is why it's weird that this error occurs.

I've attached a small sample project (which can be run using Docker) in which it's possible to reproduce this error.
[keypair_generator.zip](https://github.com/digitalbazaar/forge/files/511688/keypair_generator.zip)

Either you can run it using Docker (`docker build -t keypair_generator . && docker run --name keypair_generator keypair_generator`) or if you have Golang and Node installed you can just run the included run-script directly.

Does this help in finding the error?

**Comment by dlongley** [comment]
@Foosvald,

I ran the golang and node scripts several times without any problems. Then I modified the node script to run the `generateServerCertificate` function in a loop 100 times (with a smaller key size for speed) and bail out if it ever generated a non-PKCS#1 formatted key. It passed this test -- and given that there is no PKCS#8 code running, this is expected. So there must be something else going on that causes the key to get into PKCS#8 format.

```js
'use strict'

const fs = require('fs')
const forge = require('node-forge')

const caCert = fs.readFileSync('./ca_test.pem')
const caKey = fs.readFileSync('./ca_test.key')

/*
let cert = generateServerCertificate(
    caKey, caCert, 'etcd', ['127.0.0.1'])

fs.writeFileSync('./etcd.pem', cert.cert)
fs.writeFileSync('./etcd.key', cert.key)
*/
for(var i = 0; i < 100; ++i) {
  generateServerCertificate(
    caKey, caCert, 'etcd', ['127.0.0.1'])
}

function generateServerCertificate(caKeyString, caCertString, serviceInstance, ips) {
    var caKey = forge.pki.privateKeyFromPem(caKeyString)
    var caCert = forge.pki.certificateFromPem(caCertString)

    var keys = forge.pki.rsa.generateKeyPair(1024)
    var cert = forge.pki.createCertificate()
    cert.publicKey = keys.publicKey
    cert.serialNumber = '01'
    cert.validity.notBefore = new Date()
    cert.validity.notAfter = new Date()
    cert.validity.notAfter.setFullYear(cert.validity.notBefore.getFullYear() + 1)

    var attrsSubject = [
        {
            name: 'commonName',
            value: 'none'
        },
        {
            name: 'organizationName',
            value: 'Our Org'
        },
        {
            shortName: 'OU',
            value: serviceInstance
        }
    ]

    cert.setSubject(attrsSubject)
    cert.setIssuer(caCert.subject.attributes)

    cert.setExtensions([
        {
            name: 'subjectAltName',
            altNames: ips.map(ip => {
                return { type: 7, ip: ip }
            }),
        }])

    cert.sign(caKey, forge.md.sha256.create())

    var pem = {
        key: forge.pki.privateKeyToPem(keys.privateKey),
        cert: forge.pki.certificateToPem(cert)
    }
    if(pem.key.indexOf('BEGIN RSA PRIVATE') === -1) {
      console.log('broken key');
      fs.writeFileSync('./etcd.key', key.pem)
      process.exit(1);
    }

    return pem
}
```

**Comment by dlongley** [comment]
Btw, I've also been running the `run` script for ~10 minutes with no failures.

**Comment by Foosvald** [comment]
Hi again @dlongley,

What version of Go are you running? I get the failures in Go 1.7, but since it worked for you I tried downgrading to `go version go1.5.4 darwin/amd64`. And then everything works and Go doesn't give an error. My run script has also been running for 30 minutes now without problems.

I checked the source code of Go and they added a check for minimally encoded lengths (giving the error message `asn1: structure error: integer not minimally-encoded.`) sometime in version 1.6. https://github.com/golang/go/commit/7c20ea9311784123b72d9e45c0a29ab5cf838a3c

Is Go doing a too strict check or is Node-forge sometimes not encoding the length correctly? And do you get the problem when running Go 1.7?

Thanks again for looking into this issue!

**Comment by dlongley** [comment]
@Foosvald,

Confirmed that it can fail with Go 1.7. I'll have to look closer as to the cause when I have time.

**Comment by dlongley** [comment]
Fixed in v0.6.44. Hopefully this does not raise some esoteric bugs in other packages. If so, we'll revert and have to come up with a flag for this for 0.6.x.
